File: 2021/Java/src/main/java/com/kangjingyang/day3_2.java
package com.kangjingyang;

import java.util.ArrayList;
import java.util.LinkedList;
import java.io.*;

public class day3_2 {

    static ArrayList<String> data;
    static LinkedList<String> list;
    static int oxygen;
    static int co2;

    public static void main(String[] args) {
        data = new ArrayList<String>();
        oxygen = 0;
        co2 = 0;
        input();
        generateList();
        getOxygen();
        oxygen = Integer.parseInt(list.get(0), 2);
        generateList();
        getCo2();
        co2 = Integer.parseInt(list.get(0), 2);
        System.out.format("the ans is: %d\n", oxygen * co2);
    }

    private static void input() {
        try {
            BufferedReader in = new BufferedReader(new FileReader("../input.io"));
            StringBuffer sb;
            while (in.ready()) {
                sb = new StringBuffer(in.readLine());
                data.add(data.size(), sb.toString());
            }
            in.close();
        } catch (IOException e) {
            System.out.println("you fucked up");
        }
    }

    private static void generateList() {
        if (list != null) {
            list.clear();
        }
        list = new LinkedList<String>();
        for (int i = 0; i < data.size(); i++) {
            list.add(data.get(i));
        }
    }

    private static void getOxygen() {
        int index = 0;
        while (list.size() > 1) {
            int one = 0;
            int zero = 0;
            // count
            for (int i = 0; i < list.size(); i++) {
                if (list.get(i).charAt(index) == '0') {
                    zero++;
                } else {
                    one++;
                }
            }
            if (one >= zero) {
                // remove strings start with 0
                for (int i = 0; i < list.size(); i++) {
                    if (list.get(i).charAt(index) == '0') {
                        list.remove(i);
                        i--;
                    }
                }
            } else {
                // remove strings start with 1
                for (int i = 0; i < list.size(); i++) {
                    if (list.get(i).charAt(index) == '1') {
                        list.remove(i);
                        i--;
                    }
                }
            }
            index++;
        }
    }

    private static void getCo2() {
        int index = 0;
        while (list.size() > 1) {
            int one = 0;
            int zero = 0;
            // count
            for (int i = 0; i < list.size(); i++) {
                if (list.get(i).charAt(index) == '0') {
                    zero++;
                } else {
                    one++;
                }
            }
            if (one >= zero) {
                // remove strings start with 1
                for (int i = 0; i < list.size(); i++) {
                    if (list.get(i).charAt(index) == '1') {
                        list.remove(i);
                        i--;
                    }
                }
            } else {
                // remove strings start with 0
                for (int i = 0; i < list.size(); i++) {
                    if (list.get(i).charAt(index) == '0') {
                        list.remove(i);
                        i--;
                    }
                }
            }
            index++;
        }
    }
}

File: 2021/JS/day3-2.js
const fs = require('fs');
const lines = fs.readFileSync('../input.io', {encoding: 'utf-8'}).split('\n');

var oxygen = [[]];
var co2 = [[]];
for (let line of lines) {
    oxygen[0].push(line);
    co2[0].push(line);
}

for (let i = 0; i < lines[0].length; i++) {
    let one = 0;
    let zero = 0;
    for (let line of oxygen[oxygen.length - 1]) {
        if (line[i] === '0') {
            zero++;
        } else {
            one++;
        }
    }
    if (one === zero) {
        let temp = [];
        for (let line of oxygen[oxygen.length - 1]) {
            if (line[i] === '1') {
                temp.push(line);
            }
        }
        oxygen.push(temp);
    } else if (one + zero === 1) {
        break;
    } else if (one > zero) {
        let temp = [];
        for (let line of oxygen[oxygen.length - 1]) {
            if (line[i] === '1') {
                temp.push(line);
            }
        }
        oxygen.push(temp);
    } else {
        let temp = [];
        for (let line of oxygen[oxygen.length - 1]) {
            if (line[i] === '0') {
                temp.push(line);
            }
        }
        oxygen.push(temp);
    }
}
for (let i = 0; i < lines[0].length; i++) {
    let one = 0;
    let zero = 0;
    for (let line of co2[co2.length - 1]) {
        if (line[i] === '0') {
            zero++;
        } else {
            one++;
        }
    }
    if (one === zero) {
        let temp = [];
        for (let line of co2[co2.length - 1]) {
            if (line[i] === '0') {
                temp.push(line);
            }
        }
        console.log('temp:' + temp);
        co2.push(temp);
    } else if (one + zero === 1) {
        break;
    } else if (one < zero) {
        let temp = [];
        for (let line of co2[co2.length - 1]) {
            if (line[i] === '1') {
                temp.push(line);
            }
        }
        co2.push(temp);
    } else {
        let temp = [];
        for (let line of co2[co2.length - 1]) {
            if (line[i] === '0') {
                temp.push(line);
            }
        }
        co2.push(temp);
    }
}

console.log(
    parseInt(oxygen[oxygen.length - 1][0], 2) *
        parseInt(co2[co2.length - 1][0], 2)
);
